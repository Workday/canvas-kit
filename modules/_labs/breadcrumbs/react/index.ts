import Breadcrumbs from './lib/Breadcrumbs';

export default Breadcrumbs;
export {Breadcrumbs};
export * from './lib/Breadcrumbs';
