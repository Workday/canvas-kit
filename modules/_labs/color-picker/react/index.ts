export {default, default as ColorPicker, ColorPickerProps} from './lib/ColorPicker';
export {Swatch} from './lib/parts/Swatch';
