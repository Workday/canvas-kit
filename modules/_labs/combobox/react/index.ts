import Combobox from './lib/Combobox';

export default Combobox;
export {Combobox};
export * from './lib/Combobox';
