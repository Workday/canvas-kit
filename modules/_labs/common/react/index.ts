export * from './lib/Flex';
export * from './lib/Box';
