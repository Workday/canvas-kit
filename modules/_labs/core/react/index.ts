import type from './lib/type';
import space from './lib/space';
export default type;
export {type, space};
export * from './lib/type';
export * from './lib/StaticStates';
