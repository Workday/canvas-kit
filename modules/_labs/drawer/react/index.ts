import Drawer from './lib/Drawer';
import DrawerHeader from './lib/DrawerHeader';
export {Drawer, DrawerHeader};
export default Drawer;
export * from './lib/Drawer';
