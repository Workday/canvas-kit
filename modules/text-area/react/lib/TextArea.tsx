import * as React from 'react';
import {
  GrowthBehavior,
  ErrorType,
  errorRing,
  styled,
  Themeable,
} from '@workday/canvas-kit-react-common';
import {borderRadius, inputColors, spacingNumbers, type} from '@workday/canvas-kit-react-core';

export interface TextAreaProps
  extends Themeable,
    GrowthBehavior,
    React.TextareaHTMLAttributes<HTMLTextAreaElement> {
  /**
   * If true, set the TextArea to the disabled state.
   * @default false
   */
  disabled?: boolean;
  /**
   * The type of error associated with the TextArea (if applicable).
   */
  error?: ErrorType;
  /**
   * The ref to the inner textarea element.
   */
  inputRef?: React.Ref<HTMLTextAreaElement>;
  /**
   * The function called when the TextArea state changes.
   */
  onChange?: (e: React.ChangeEvent<HTMLTextAreaElement>) => void;
  /**
   * The placeholder text of the TextArea.
   */
  placeholder?: string;
  /**
   * If true, set the TextArea to read-only. The user will be unable to interact with the TextArea.
   * @default false
   */
  readOnly?: boolean;
  /**
   * The resize constraints of the TextArea.
   * @default TextArea.ResizeDirection.Both
   */
  resize?: TextAreaResizeDirection;
  /**
   * The value of the TextArea.
   */
  value?: any;
}

export enum TextAreaResizeDirection {
  None = 'none',
  Both = 'both',
  Horizontal = 'horizontal',
  Vertical = 'vertical',
}

const TextAreaContainer = styled('textarea')<TextAreaProps>(
  ({theme, error}) => ({
    ...type.body,
    border: `1px solid ${inputColors.border}`,
    display: 'block',
    backgroundColor: inputColors.background,
    borderRadius: borderRadius.m,
    boxSizing: 'border-box',
    minHeight: 64,
    minWidth: 280,
    transition: '0.2s box-shadow, 0.2s border-color',
    padding: spacingNumbers.xxs, // Compensate for border
    margin: 0, // Fix Safari
    '&::webkit-resizer': {
      display: 'none',
    },
    '&::placeholder': {
      color: inputColors.placeholder,
    },
    '&:hover': {
      borderColor: inputColors.hoverBorder,
    },
    '&:focus:not([disabled])': {
      borderColor: theme.palette.common.focusOutline,
      boxShadow: `inset 0 0 0 1px ${theme.palette.common.focusOutline}`,
      outline: 'none',
    },
    '&:disabled': {
      backgroundColor: inputColors.disabled.background,
      borderColor: inputColors.disabled.border,
      color: inputColors.disabled.text,
      '&::placeholder': {
        color: inputColors.disabled.text,
      },
    },
    ...errorRing(error, theme),
  }),

  ({resize, grow}) => ({
    width: grow ? '100%' : undefined,
    resize: grow ? TextAreaResizeDirection.Vertical : resize,
  })
);

export default class TextArea extends React.Component<TextAreaProps> {
  static ErrorType = ErrorType;
  static ResizeDirection = TextAreaResizeDirection;

  render() {
    const {resize = TextAreaResizeDirection.Both, grow, inputRef, ...inputProps} = this.props;

    return <TextAreaContainer ref={inputRef} grow={grow} resize={resize} {...inputProps} />;
  }
}
