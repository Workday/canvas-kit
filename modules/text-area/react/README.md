# Canvas Kit Text Area

Canvas-styled React textarea component.

## Installation

```sh
yarn add @workday/canvas-kit-react
```

or

```sh
yarn add @workday/canvas-kit-react-text-area
```

# TextArea

#### Simple Example

**Note:** While a base textarea component is provided in this package, it is **not accessible** when
used as is. It should be used in tandem with [`FormField`](../../form-field/react) to be made fully
accessible (see below).

```tsx
import * as React from 'react';
import TextArea from '@workday/canvas-kit-react-text-area';

<TextArea placeholder="Placeholder" value={this.state.value} onChange={() => {}} />;
```

#### Accessible Example

```tsx
import * as React from 'react';
import TextArea from '@workday/canvas-kit-react-text-area';
import FormField from '@workday/canvas-kit-react-form-field';

// Accessible default state
<FormField label="My Field" inputId="my-textarea-field">
  <TextArea placeholder="Placeholder" value={this.state.value} onChange={() => {}} />;
</FormField>

// Accessible error state
<FormField
  label="My Field"
  inputId="my-textarea-field"
  hintText="Helpful text to resolve error"
  hintId="my-textarea-field-error"
  error={FormField.ErrorType.Error}
>
  <TextArea placeholder="Placeholder" value={this.state.value} onChange={() => {}} />;
</FormField>
```

If use inside a FormField doesn't work for your use case, you can use the `aria-labelledby`
attribute.

```tsx
import * as React from 'react';
import TextArea from '@workday/canvas-kit-react-text-area';
<label id="123">Label</label>
...
<TextArea placeholder="Placeholder" value={this.state.value} onChange={() => {}} aria-labelledby="123" />;
```

## Static Properties

#### `ErrorType: ErrorType`

```tsx
<TextArea error={TextArea.ErrorType.Alert} />
```

## Component Props

### Required

None

### Optional

#### `disabled: boolean`

> Whether or not the text area is disabled.

Default: `false`

---

#### `error: ErrorType`

> The type of error to display, if any.

| Type  | Description                     |
| ----- | ------------------------------- |
| Error | Red outline with error icon.    |
| Alert | Yellow outline with alert icon. |

---

#### `inputRef: React.Ref<HTMLTextAreaElement>`

> The ref callback for the inner text area element.

---

#### `onChange: React.ChangeEventHandler<HTMLTextAreaElement>`

> The callback fired when the value is changed.

---

#### `placeholder: string`

> Placeholder text to be displayed if there is no input value.

Default: `''`

---

#### `readOnly: boolean`

> If true, user will be unable to interact with the field.

Default: `false`

---

#### `value: any`

> Text Area value

---

#### `resize: TextAreaResizeDirection`

> Set resize constraints on the text area

| Direction                           |
| ----------------------------------- |
| TextArea.ResizeDirection.Both       |
| TextArea.ResizeDirection.None       |
| TextArea.ResizeDirection.Vertical   |
| TextArea.ResizeDirection.Horizontal |

Default: `TextArea.ResizeDirection.Both`

---
