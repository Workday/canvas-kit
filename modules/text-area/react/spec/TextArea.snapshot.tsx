import * as React from 'react';
import * as renderer from 'react-test-renderer';
import TextArea from '../lib/TextArea';

describe('TextArea Snapshots', () => {
  test('renders as expected', () => {
    const component = renderer.create(<TextArea />);
    expect(component).toMatchSnapshot();
  });

  test('grow', () => {
    const component = renderer.create(<TextArea grow={true} />);
    expect(component).toMatchSnapshot();
  });

  test('error', () => {
    const component = renderer.create(<TextArea error={TextArea.ErrorType.Error} />);
    expect(component).toMatchSnapshot();
  });
});
