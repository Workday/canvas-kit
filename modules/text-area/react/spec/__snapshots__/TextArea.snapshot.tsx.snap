// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea Snapshots error 1`] = `
.emotion-0 {
  font-size: 14px;
  line-height: 20px;
  font-weight: 400;
  color: #494949;
  font-family: "Roboto","Helvetica Neue","Helvetica",Arial,sans-serif;
  border: 1px solid #7b858f;
  display: block;
  background-color: #ffffff;
  border-radius: 4px;
  box-sizing: border-box;
  min-height: 64px;
  min-width: 280px;
  -webkit-transition: 0.2s box-shadow,0.2s border-color;
  transition: 0.2s box-shadow,0.2s border-color;
  padding: 8px;
  border-color: #de2e21;
  -webkit-transition: 100ms box-shadow;
  transition: 100ms box-shadow;
  box-shadow: inset 0 0 0 1px #de2e21;
  resize: both;
}

.emotion-0::webkit-resizer {
  display: none;
}

.emotion-0::-webkit-input-placeholder {
  color: #5e6a75;
}

.emotion-0::-moz-placeholder {
  color: #5e6a75;
}

.emotion-0:-ms-input-placeholder {
  color: #5e6a75;
}

.emotion-0::placeholder {
  color: #5e6a75;
}

.emotion-0:focus:not([disabled]) {
  border-color: #0875e1;
  box-shadow: inset 0 0 0 1px #0875e1;
  outline: none;
}

.emotion-0:disabled {
  background-color: #f6f7f8;
  border-color: #b9c0c7;
  color: #a1aab3;
}

.emotion-0:disabled::-webkit-input-placeholder {
  color: #a1aab3;
}

.emotion-0:disabled::-moz-placeholder {
  color: #a1aab3;
}

.emotion-0:disabled:-ms-input-placeholder {
  color: #a1aab3;
}

.emotion-0:disabled::placeholder {
  color: #a1aab3;
}

.emotion-0:hover {
  border-color: #de2e21;
}

.emotion-0:focus:not([disabled]) {
  border-color: #de2e21;
  box-shadow: inset 0 0 0 1px #de2e21, 0 0 0 2px #ffffff, 0 0 0 4px #0875e1;
}

<textarea
  className="emotion-0 emotion-1"
/>
`;

exports[`TextArea Snapshots grow 1`] = `
.emotion-0 {
  font-size: 14px;
  line-height: 20px;
  font-weight: 400;
  color: #494949;
  font-family: "Roboto","Helvetica Neue","Helvetica",Arial,sans-serif;
  border: 1px solid #7b858f;
  display: block;
  background-color: #ffffff;
  border-radius: 4px;
  box-sizing: border-box;
  min-height: 64px;
  min-width: 280px;
  -webkit-transition: 0.2s box-shadow,0.2s border-color;
  transition: 0.2s box-shadow,0.2s border-color;
  padding: 8px;
  width: 100%;
  resize: vertical;
}

.emotion-0::webkit-resizer {
  display: none;
}

.emotion-0::-webkit-input-placeholder {
  color: #5e6a75;
}

.emotion-0::-moz-placeholder {
  color: #5e6a75;
}

.emotion-0:-ms-input-placeholder {
  color: #5e6a75;
}

.emotion-0::placeholder {
  color: #5e6a75;
}

.emotion-0:focus:not([disabled]) {
  border-color: #0875e1;
  box-shadow: inset 0 0 0 1px #0875e1;
  outline: none;
}

.emotion-0:disabled {
  background-color: #f6f7f8;
  border-color: #b9c0c7;
  color: #a1aab3;
}

.emotion-0:disabled::-webkit-input-placeholder {
  color: #a1aab3;
}

.emotion-0:disabled::-moz-placeholder {
  color: #a1aab3;
}

.emotion-0:disabled:-ms-input-placeholder {
  color: #a1aab3;
}

.emotion-0:disabled::placeholder {
  color: #a1aab3;
}

<textarea
  className="emotion-0 emotion-1"
/>
`;

exports[`TextArea Snapshots renders as expected 1`] = `
.emotion-0 {
  font-size: 14px;
  line-height: 20px;
  font-weight: 400;
  color: #494949;
  font-family: "Roboto","Helvetica Neue","Helvetica",Arial,sans-serif;
  border: 1px solid #7b858f;
  display: block;
  background-color: #ffffff;
  border-radius: 4px;
  box-sizing: border-box;
  min-height: 64px;
  min-width: 280px;
  -webkit-transition: 0.2s box-shadow,0.2s border-color;
  transition: 0.2s box-shadow,0.2s border-color;
  padding: 8px;
  resize: both;
}

.emotion-0::webkit-resizer {
  display: none;
}

.emotion-0::-webkit-input-placeholder {
  color: #5e6a75;
}

.emotion-0::-moz-placeholder {
  color: #5e6a75;
}

.emotion-0:-ms-input-placeholder {
  color: #5e6a75;
}

.emotion-0::placeholder {
  color: #5e6a75;
}

.emotion-0:focus:not([disabled]) {
  border-color: #0875e1;
  box-shadow: inset 0 0 0 1px #0875e1;
  outline: none;
}

.emotion-0:disabled {
  background-color: #f6f7f8;
  border-color: #b9c0c7;
  color: #a1aab3;
}

.emotion-0:disabled::-webkit-input-placeholder {
  color: #a1aab3;
}

.emotion-0:disabled::-moz-placeholder {
  color: #a1aab3;
}

.emotion-0:disabled:-ms-input-placeholder {
  color: #a1aab3;
}

.emotion-0:disabled::placeholder {
  color: #a1aab3;
}

<textarea
  className="emotion-0 emotion-1"
/>
`;
